### PR TITLE
[Attention] Add MLA prefill backend: trtllm_ragged_attention_deepseek

### PR DESCRIPTION
**Pull request by minosfuture**
## Purpose

Add MLA prefill backend: trtllm_ragged_attention_deepseek
* controlled by `VLLM_USE_TRTLLM_RAGGED_DEEPSEEK_PREFILL=1`

* [x] Need to fix accuracy issue
  * 0.54 -> 0.59 fixed by zeroing output tensor for context chunk prefill
  * 0.59 -> 0.61 (0.22 -> 0.64 w/ prefix caching) fixed by transposing lse
  * 0.61 -> 0.65 fixed by zeroing workspace
* [x] performance benchmark

## Test Plan

* [ ] microbenchmark
  * in follow-up work after refactoring
* [x] vllm bench serve
* [x] lm_eval gsm8k

### bench serve

12.7% qps improvement
-10.2% TTFT improvement

#### Baseline
```
CUDA_VISIBLE_DEVICES=0 VLLM_LOGGING_LEVEL=DEBUG VLLM_ATTENTION_BACKEND=FLASHINFER_MLA vllm serve --model="deepseek-ai/DeepSeek-V2-Lite-Chat" 2>&1 | tee /tmp/fi.log
```
```
============ Serving Benchmark Result ============
Successful requests:                     2048
Maximum request concurrency:             2048
Benchmark duration (s):                  200.42
Total input tokens:                      4192256
Total generated tokens:                  2097152
Request throughput (req/s):              10.22
Output token throughput (tok/s):         10463.54
Peak output token throughput (tok/s):    16094.00
Peak concurrent requests:                2048.00
Total Token throughput (tok/s):          31380.41
---------------Time to First Token----------------
Mean TTFT (ms):                          64511.82
Median TTFT (ms):                        55968.36
P99 TTFT (ms):                           132592.32
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          86.54
Median TPOT (ms):                        89.57
P99 TPOT (ms):                           93.98
---------------Inter-token Latency----------------
Mean ITL (ms):                           86.54
Median ITL (ms):                         70.26
P99 ITL (ms):                            208.33
==================================================
```
#### This PR
```
CUDA_VISIBLE_DEVICES=0 VLLM_LOGGING_LEVEL=DEBUG VLLM_ATTENTION_BACKEND=FLASHINFER_MLA VLLM_USE_TRTLLM_RAGGED_DEEPSEEK_PREFILL=1 vllm serve --model="deepseek-ai/DeepSeek-V2-Lite-Chat" 2>&1 | tee /tmp/ragged.log
```

```
============ Serving Benchmark Result ============
Successful requests:                     2048
Maximum request concurrency:             2048
Benchmark duration (s):                  177.74
Total input tokens:                      4192256
Total generated tokens:                  2097152
Request throughput (req/s):              11.52
Output token throughput (tok/s):         11799.21
Peak output token throughput (tok/s):    18098.00
Peak concurrent requests:                2048.00
Total Token throughput (tok/s):          35386.10
---------------Time to First Token----------------
Mean TTFT (ms):                          57924.67
Median TTFT (ms):                        49533.53
P99 TTFT (ms):                           118626.18
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          76.15
Median TPOT (ms):                        78.87
P99 TPOT (ms):                           83.38
---------------Inter-token Latency----------------
Mean ITL (ms):                           76.16
Median ITL (ms):                         60.18
P99 ITL (ms):                            181.91
==================================================
```

#### Baseline
```
CUDA_VISIBLE_DEVICES=0 VLLM_LOGGING_LEVEL=DEBUG VLLM_ATTENTION_BACKEND=FLASHINFER_MLA vllm serve --model="deepseek-ai/DeepSeek-V2-Lite-Chat" --enforce-eager 2>&1 | tee /tmp/fi.log

local-completions (model=deepseek-ai/DeepSeek-V2-Lite-Chat,base_url=http://127.0.0.1:8000/v1/completions,num_concurrent=256), gen_kwargs: (None), limit: None, num_fewshot: 8, batch_size: 1
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     8|exact_match|↑  |0.6520|±  |0.0131|
|     |       |strict-match    |     8|exact_match|↑  |0.6482|±  |0.0132|

#### This PR
```
CUDA_VISIBLE_DEVICES=0 VLLM_LOGGING_LEVEL=DEBUG VLLM_ATTENTION_BACKEND=FLASHINFER_MLA VLLM_USE_TRTLLM_RAGGED_DEEPSEEK_PREFILL=1 vllm serve --model="deepseek-ai/DeepSeek-V2-Lite-Chat" --enforce-eager 2>&1 | tee /tmp/ragged.log

local-completions (model=deepseek-ai/DeepSeek-V2-Lite-Chat,base_url=http://127.0.0.1:8000/v1/completions,num_concurrent=256), gen_kwargs: (None), limit: None, num_fewshot: 8, batch_size: 1
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     8|exact_match|↑  |0.655|±  |0.0131|
|     |       |strict-match    |     8|exact_match|↑  |0.649|±  |0.0131|

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/v1/attention/backends/mla/common.py [hunk 1]
```diff
@@ -1370,6 +1393,64 @@ def _run_prefill_context_chunk_cudnn(
...
+            max_q_len=prefill.max_query_len,
+            max_kv_len=prefill.max_query_len,
+            bmm1_scale=self.scale,
+            bmm2_scale=1.0,
+            o_sf_scale=1.0,
+            batch_size=prefill.query_seq_lens.shape[0],
+            window_left=-1,
+            cum_seq_lens_q=prefill.query_start_loc,
+            cum_seq_lens_kv=prefill.query_start_loc,
+            enable_pdl=False,
```
Comment (minosfuture) [inline vllm/v1/attention/backends/mla/common.py:1474]: todo: check if pdl is supported

File: vllm/v1/attention/backends/mla/common.py [hunk 2]
```diff
@@ -613,6 +624,11 @@ def __init__(
                 get_per_layer_parameters(vllm_config, layer_names, MLACommonImpl)
             )
 
+        if self._use_trtllm_ragged_prefill:
```
Comment (pavanimajety) [inline vllm/v1/attention/backends/mla/common.py:627]: Does it make sense to add the `trtllm_ragged_prefill` as the prefill kernel for flashinfer_mla.py file/backend?

**Review by LucasWilkinson — APPROVED** [review]
LGTM :+1:; Thanks for the contribution; would be nice for @pavanimajety to also take look 👍

File: vllm/v1/attention/backends/mla/common.py (hunk 1 above)
Comment (pavanimajety) [reply vllm/v1/attention/backends/mla/common.py:1474]: Checked with the kernel author - they say PDL is supported. Do we need additional testing with enable_pdl=True?

File: vllm/v1/attention/backends/mla/common.py (hunk 2 above)
Comment (minosfuture) [reply vllm/v1/attention/backends/mla/common.py:627]: This PR follows the current prefill kernel integration pattern. But I do think we need to refactor the MLA prefill backends here. Currently they are all put in the common.py, and it would be better to organize them into separate files.
Let's do this refactoring in a follow-up PR, and keep the integration style consistent for now?

Comment (pavanimajety) [reply vllm/v1/attention/backends/mla/common.py:627]: works for me, thanks!

**Review by pavanimajety — APPROVED** [review]

File: vllm/v1/attention/backends/mla/common.py [hunk 3]
```diff
@@ -1418,6 +1448,81 @@ def _run_prefill_context_chunk_cudnn(
...
+    def _run_prefill_context_chunk_trtllm_ragged(
+        self, prefill: MLACommonPrefillMetadata, chunk_idx: int, q, k, v
+    ):
+        """TRT-LLM ragged attention for context chunks (non-causal)."""
+        from flashinfer.prefill import trtllm_ragged_attention_deepseek
+
+        assert prefill.chunked_context is not None
+        assert prefill.chunked_context.seq_lens[chunk_idx] is not None
+
+        out = torch.zeros(
```
Comment (robertgshaw2-redhat) [inline vllm/v1/attention/backends/mla/common.py:1493]: @minosfuture do you know why its needed to fill these with 0s?

I noticed that creating these 0's tensors takes non-trivial time on the hotpath

Comment (jdebache) [reply vllm/v1/attention/backends/mla/common.py:1493]: We are on it :)
